tests/python_package_test/test_engine.py

```diff
@@ -1545,6 +1545,35 @@ def test_refit():
     assert err_pred > new_err_pred
 
 
+def test_refit_dataset_params():
+    # check refit accepts dataset_params
+    X, y = load_breast_cancer(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+    lgb_train = lgb.Dataset(X_train, y_train)
+    train_params = {
+        'objective': 'binary',
+        'metric': 'binary_logloss',
+        'verbose': -1,
+        'min_data': 10,
+        'seed': 123
+    }
+    gbm = lgb.train(train_params, lgb_train, num_boost_round=20)
+    non_weight_err_pred = log_loss(y_test, gbm.predict(X_test))
+    dataset_params = {
+        'max_bin': 260,
+        'min_data_in_bin': 5,
+        'data_random_seed': 123,
+    }
+    new_gbm = gbm.refit(
+        data=X_train,
+        label=y_train,
+        weight=np.random.rand(y_train.shape[0]),
+        dataset_params=dataset_params,
+    )
+    weight_err_pred = log_loss(y_test, new_gbm.predict(X_test))
+    assert weight_err_pred != non_weight_err_pred
+
```

Comment:
```suggestion
    train_set_params = new_gbm.train_set.get_params()
    assert train_set_params["max_bin"] == 260
    assert train_set_params["min_data_in_bin"] == 5
    assert train_set_params["data_random_seed"] == 123
```

Since you are specifying `Dataset` parameters here, can you please test that they're passed through correctly?

Adding such tests will increase confidence that this change is working as expected.

Reply:
And actually, now that I think about it...could you also directly extract the weight and compare it to the one that was passed in?

Like

```python
refit_weight = np.random.rand(y_train.shape[0])
new_gbm = gbm.refit(
        data=X_train,
        label=y_train,
        weight=refit_weight,
        dataset_params=dataset_params,
)
stored_weights = new_gbm.train_set.get_weight()
assert stored_weights == refit_weight
```

Reply:
here I'll use [np.assert_all_close](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) because tests array of float numbers, or do you have another thing in mind?

Reply:
yep, sounds good!

Reply:
@jameslamb ci is green, let me know if anything else is needed